keps/sig-node/3759-sidecar-containers/README.md

```diff
@@ -0,0 +1,1235 @@
+effectively become a sidecar container similar to Init container with the same
+field value.
+
+Additional considerations can be found in the older document that was a previous
+attempt to implement the sidecar containers:
+https://docs.google.com/document/d/1gX_SOZXCNMcIe9d8CkekiT0GU9htH_my8OB7HSm3fM8/edit#heading=h.71o8fcesvgba.
+
+### Naming
+
+#### Collection name
+
+For this KEP it is important to have sidecar containers be defined among other
+Init containers to be able to express the initialization order of containers.
+The name `initContainers` is not a good fit for sidecar containers as they
+typically do more than initialization. The better name would be “infrastructure”
+containers. The current idea is to implement sidecars as a part of init
+containers and if this introduces too much trouble, the new collection name may
+replace the old collection name in future. 
+
+#### Reuse of restartPolicy field and enum
+
+The per-container restart policy was a long standing request from the community.
+Implementing per-container restart policy introduces a set of challenging
+problems for the pod lifecycle. For example, the state keeping for containers
+which has already run to completion. Introducing sidecar containers is another
+scenario where this field can be semantically used.
+
+Introducing this field on containers opens up opportunities to implement those
+long-standing requests from the community.
+
+#### Use “Always” vs. New enum value
+
+The semantic of an “Always” enum value is very close to what sidecars require.
+This is why reusing Always to represent sidecars makes a lot of sense for Init
+containers.
+
+Why “Always” that may not look similar to Always for other containers makes
+sense:
+
+Pros:
+- Allows the same enum values to be used for the `restartPolicy` of both pods
+and containers, but semantics that are mostly the same. Cons:
+- There are slight differences between the semantics of `Always` for containers
+and `Always` for pods. The main difference is that a initContainer with `Always`
+will be terminated when the pod terminates and has no influence over the pod
+lifecycle. Also, for Pods, the default `restartPolicy` is
+[documented](https://github.com/kubernetes/kubernetes/blob/280473ebc4e45f9001f5f9789c318ff7329bc5f0/staging/src/k8s.io/api/core/v1/types.go#L2753-L2764)
+as `Always` but for initContainers it will default to `OnFailure`. We believe
+this can be addressed in the documentation of the existing enum fields.
+
+When in future we will support “Always” on regular containers, there will be
+interesting case of “Always” having a meaning of non-blocking container for the
+Pod with restartPolicy “Never” or “OnFailure”. This is easy to explain - Pod
+lifecycle is controlled by containers with the matching restartPolicy. But it
+may be slightly confusing.
+
+  
+#### Alternatives
+
+- Introduce a new collection `infrastructureContainers` that replaces
```

Comment:
I don't like this because it removes the possibility of ordering init containers and infra containers, and since there ARE use cases where infra containers need init containers, I don't want to have cross ordering, so that is a very strong argument to me that we should keep them both as types of init containers.

Reply:
I reworded. This alternatvie was about renaming the entire `initContainers` section in favor of a new name. Alternative with the sidecar containers as a separate collection is listed in the end of the document as rejected